store/tikv/tikvrpc/tikvrpc.go

```diff
@@ -354,3 +373,142 @@ func (resp *Response) GetRegionError() (*errorpb.Error, error) {
+	case CmdCommit:
+		resp.Commit, err = client.KvCommit(ctx, req.Commit)
+	case CmdCleanup:
+		resp.Cleanup, err = client.KvCleanup(ctx, req.Cleanup)
+	case CmdBatchGet:
+		resp.BatchGet, err = client.KvBatchGet(ctx, req.BatchGet)
+	case CmdBatchRollback:
+		resp.BatchRollback, err = client.KvBatchRollback(ctx, req.BatchRollback)
+	case CmdScanLock:
+		resp.ScanLock, err = client.KvScanLock(ctx, req.ScanLock)
+	case CmdResolveLock:
+		resp.ResolveLock, err = client.KvResolveLock(ctx, req.ResolveLock)
+	case CmdGC:
+		resp.GC, err = client.KvGC(ctx, req.GC)
+	case CmdDeleteRange:
+		resp.DeleteRange, err = client.KvDeleteRange(ctx, req.DeleteRange)
+	case CmdRawGet:
+		resp.RawGet, err = client.RawGet(ctx, req.RawGet)
+	case CmdRawPut:
+		resp.RawPut, err = client.RawPut(ctx, req.RawPut)
+	case CmdRawDelete:
+		resp.RawDelete, err = client.RawDelete(ctx, req.RawDelete)
+	case CmdRawScan:
+		resp.RawScan, err = client.RawScan(ctx, req.RawScan)
+	case CmdCop:
+		resp.Cop, err = client.Coprocessor(ctx, req.Cop)
+	case CmdCopStream:
+		// Use context to support timeout for grpc streaming client.
+		ctx1, cancel := context.WithCancel(ctx)
+		var stream tikvpb.Tikv_CoprocessorStreamClient
+		stream, err = client.CoprocessorStream(ctx1, req.Cop)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		// Read the first streaming response to get CopStreamResponse.
+		// This can make error handling much easier, because SendReq() retry on
+		// region error automatically.
+		var first *coprocessor.Response
+		first, err = resp.CopStream.Recv()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		resp.CopStream = NewCopStreamResponse(stream, first, cancel, ch)
+	case CmdMvccGetByKey:
+		resp.MvccGetByKey, err = client.MvccGetByKey(ctx, req.MvccGetByKey)
+	case CmdMvccGetByStartTs:
+		resp.MvccGetByStartTS, err = client.MvccGetByStartTs(ctx, req.MvccGetByStartTs)
+	case CmdSplitRegion:
+		resp.SplitRegion, err = client.SplitRegion(ctx, req.SplitRegion)
+	default:
+		return nil, errors.Errorf("invalid request type: %v", req.Type)
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return resp, nil
+}
+
+// TimeoutItem is used to implement grpc stream timeout.
+type TimeoutItem struct {
+	cancel  context.CancelFunc
+	timeout time.Time
+	active  int32 // atomic variable, 0 means inactive
+}
+
+// SetDeadline set deadline for this CopStreamResponse object, it usually called before Recv().
+func (resp *CopStreamResponse) SetDeadline(deadline time.Time) {
+	item := &TimeoutItem{
+		cancel:  resp.cancel,
+		timeout: deadline,
+		active:  1,
+	}
+	resp.pending = item
+	resp.timeoutCh <- item
+}
+
+// Recv overrides the stream client Recv() function.
+func (resp *CopStreamResponse) Recv() (*coprocessor.Response, error) {
+	ret, err := resp.Tikv_CoprocessorStreamClient.Recv()
+	if resp.pending != nil {
+		atomic.StoreInt32(&resp.pending.active, 0)
+		resp.pending = nil
+	}
+	return ret, errors.Trace(err)
+}
```

Comment:
If there is always a `SetDeadline()` before `Recv()`, seems better to merge them to one method.